synapse/storage/databases/main/relations.py

```diff
@@ -153,6 +153,12 @@ def _get_recent_references_for_event_txn(
             # If there are more events, generate the next pagination key.
             next_token = None
             if len(events) > limit and last_topo_id and last_stream_id:
+                # Due to how the pagination clause is generated, the stream ID is
+                # handled as an exclusive range when paginating forward. Correct
+                # for that here.
```

Comment:
Does this refer to the `limit + 1` on line 140? Presumably we test to see if we got the `+1` event to work out if there's more to paginate.

Not sure about the phrase `handled as an exclusive range`. Perhaps `is an exclusive upper bound`? :shrug:

Reply:
I'm somewhat trying to match code used in the `/messages` API:

https://github.com/matrix-org/synapse/blob/d3d9ca156e323fe194b1bcb1af1628f65a2f3c1c/synapse/storage/databases/main/stream.py#L1335-L1345

I'm a bit confused though at why that one applies when paginating backwards and this one applies when paginating forwards, empirically this fixes the issue though. I'm not 100% clear what's going on.

Reply:
Oooo I think I spotted a difference in implementation, I'm going to try to abstract some code.

Reply:
I ended up porting the code over from `_paginate_room_events_txn` -- essentially this was an off-by-one error. See 8ce01fbfcd976c541ba3f61e2a6eab8e55b136f2. Instead of looking at the *last* row (which is what tells us there's more data), we want to look at the second to last row.